provider/provider.go

```diff
@@ -138,11 +213,92 @@ func LoadFromDisk(homedir string) (*Provider, error) {
 	return provider, err
 }
 
-func (provider *Provider) addActivity(activity Activity) {
-	provider.activity = append(provider.activity, activity)
-	if len(provider.activity) > maxActivity {
-		// Drop the oldest activity.
-		provider.activity = provider.activity[1:]
+// Add statistics in Time Series format for charting
+func (provider *Provider) addActivity(op string, bytes int64) {
+	metrics := []*Activity{&provider.stats.Hour, &provider.stats.Day, &provider.stats.Week}
+	for _, act := range metrics {
+		// add and cycle activity within metric as needed
+		provider.cycleActivity(act)
+
+		// update data in current frame
+		idx := len(act.Timestamps) - 1
+		if op == "upload" {
+			act.BlockUploads[idx]++
+			act.BytesUploaded[idx] += bytes
+		}
+		if op == "download" {
+			act.BlockDownloads[idx]++
+			act.BytesDownloaded[idx] += bytes
+		}
+		if op == "delete" {
+			act.BlockDeletions[idx]++
+		}
+		if op == "contract" {
+			act.StorageReservations[idx]++
+		}
+	}
+	// update provider information
+	if op == "upload" {
+		provider.TotalBlocks++
+		provider.StorageUsed += bytes
+	}
+	if op == "delete" {
+		provider.TotalBlocks--
+		provider.StorageUsed -= bytes
+	}
+	if op == "contract" {
+		provider.StorageReserved += bytes
+	}
+
+}
+
+// takes in a pointer to an activity cycle and adds any new activity cycles
+// and discards old cycles
+func (provider *Provider) cycleActivity(act *Activity) {
+	provider.mu.Lock()
+	defer provider.mu.Unlock()
```

Comment:
I might lift the locking up a function or two. This won't protect the stats since they're mutated by the caller without the lock held anyways.